File: OdontoSysV2/OdontoSys/OdontoSysBussiness/src/main/java/com/odontosys/bo/CitaBO.java
package com.odontosys.bo;

import com.odontosys.dao.services.CitaDAO;
import com.odontosys.daoImp.services.CitaDAOImpl;
import com.odontosys.infrastructure.model.Turno;
import com.odontosys.services.model.Cita;
import com.odontosys.users.model.Odontologo;
import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;

public class CitaBO {
    private CitaDAO citaDAO;
    private PacienteBO pacienteBO;
    
    public CitaBO(){
        this.citaDAO = new CitaDAOImpl();
        this.pacienteBO = new PacienteBO();
    }
    
    public Integer insertCita(Cita cita){
        return this.citaDAO.insertar(cita);
    }
    
    public Integer modificarCita(Cita cita){
        return this.citaDAO.modificar(cita);
    }
    
    public Integer eliminarCita(Cita cita){
        return this.citaDAO.eliminar(cita);
    }
    
    public ArrayList<Cita>listarPorOdontologo(Odontologo odontologo,String fechaInicio,String fechaFin){
        ArrayList<Cita>lista = this.citaDAO.listarPorOdontologo(odontologo,LocalDate.parse(fechaInicio),LocalDate.parse(fechaFin));
        return lista;
    }
    
    //Magia de @jared
    public boolean[][] calcularDisponibilidad(ArrayList<Cita> citas, ArrayList<Turno> turnos, String fechaInicio){
        boolean[][] disponibilidad = new boolean[7][48];
        LocalDate inicio = LocalDate.parse(fechaInicio);
        
        for(int i=0; i<7; i++){
            LocalDate fecha = inicio.plusDays(i);
            DayOfWeek dia = fecha.getDayOfWeek();
            
            for(Turno turno : turnos){
                if(turno.getDiaSemana().toString().equals(dia.toString())){
                    int inicioBloque = (turno.getHoraInicio().getHour() * 2) + 
                                      (turno.getHoraInicio().getMinute() >= 30 ? 1 : 0);
                    int finBloque = (turno.getHoraFin().getHour() * 2) + 
                                   (turno.getHoraFin().getMinute() > 0 ? 1 : 0);
                    
                    for(int j = inicioBloque; j < finBloque; j++)
                        disponibilidad[i][j] = true;
                }
            }
            
            for (Cita cita : citas){
                LocalDate fechaCita = cita.getFecha();
                int diferencia = (int) ChronoUnit.DAYS.between(inicio, fechaCita);
                if(diferencia >= 0 && diferencia < 7){
                    int bloque = cita.getHoraInicio().getHour() * 2 + 
                            (cita.getHoraInicio().getMinute() >= 30 ? 1 : 0);
                    disponibilidad[diferencia][bloque] = false;
                }
            }
        }
        return disponibilidad;
    }

}

File: OdontoSysFrontEnd/OdontoSysBusiness/BO/CitaBO.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;
using OdontoSysBusiness.citaWS;

namespace OdontoSysBusiness.BO
{
    public class CitaBO
    {
        private CitasClient citasClienteSOAP;

        public CitaBO() 
        {
            this.citasClienteSOAP = new CitasClient();
        }

        public int InsertCita(cita cita)
        {
            return this.citasClienteSOAP.insertCita(cita);
        }

        public int ModificarCita(cita cita)
        {
            return this.citasClienteSOAP.modificarCita(cita);
        }

        public int EliminarCita(cita cita)
        {
            return this.citasClienteSOAP.eliminarCita(cita);
        }

        public BindingList<cita>ListarPorOdontologo(odontologo odontologo, string fechaInicio, string fechaFin)
        {
            cita[] lista = citasClienteSOAP.od_listarPorOdontologo(odontologo, fechaInicio, fechaFin);
            return new BindingList<cita>(lista);

        }

        public booleanArray[] CalcularDisponibilidad(cita[] citas, turno[] turnos, string fechaInicio)
        { 
            return this.citasClienteSOAP.calcularDisponibilidad(citas, turnos, fechaInicio);
        }
    }
}
